x/mongo/driver/topology/server.go

```diff
@@ -707,28 +725,34 @@ func (s *Server) check() (description.Server, error) {
 	var err error
 	var durationNanos int64
 
+	previousDescription := s.Description()
+	streamable := previousDescription.TopologyVersion != nil
+
+	isNilConn := s.conn == nil
+	if !isNilConn {
+		s.publishServerHeartbeatStartedEvent(s.conn.ID(), s.conn.getCurrentlyStreaming() || streamable)
+	}
 	// Create a new connection if this is the first check, the connection was closed after an error during the previous
 	// check, or the previous check was cancelled.
-	if s.conn == nil || s.conn.closed() || s.checkWasCancelled() {
+	if isNilConn || s.conn.closed() || s.checkWasCancelled() {
 		// Create a new connection and add it's handshake RTT as a sample.
 		err = s.setupHeartbeatConnection()
 		if err == nil {
 			// Use the description from the connection handshake as the value for this check.
 			s.rttMonitor.addSample(s.conn.helloRTT)
 			descPtr = &s.conn.desc
+		} else {
+			err = unwrapConnectionError(err)
```

Comment:
Is there a reason we need to unwrap the connection error here? Can the unwrapped connection error still provide a topology version via `extractTopologyVersion` on line 819?

Reply:
> I unwrap the error here so we can use a uniform logic in update() to detect timeout errors.

Don't we call `unwrapConnectionError` in `update()` already [here](https://github.com/mongodb/mongo-go-driver/pull/1133/files#diff-5e467b910a8117877d35340d0673292dbe5435fb482aa701363a22984daec72dR559)?

Reply:
The unwrapped error will be wrapped again at l830 by `description.NewServerFromError()`, which is then unwrapped by `unwrapConnectionError` in `update()`.